src/UndockedRegFreeWinRT/UndockedRegFreeWinRT/catalog.cpp

```diff
@@ -81,72 +82,373 @@ struct component
             {
-                auto this_component = make_shared<component>();
-                this_component->module_name = fileName;
-
-                const WCHAR* threadingModel;
-                RETURN_IF_FAILED(xmlReader->MoveToAttributeByName(L"threadingModel", nullptr));
-                RETURN_IF_FAILED(xmlReader->GetValue(&threadingModel, nullptr));
-
-                if (_wcsicmp_l(L"sta", threadingModel, locale) == 0)
-                {
-                    this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_STA;
-                }
-                else if (_wcsicmp_l(L"mta", threadingModel, locale) == 0)
-                {
-                    this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_MTA;
-                }
-                else if (_wcsicmp_l(L"both", threadingModel, locale) == 0)
-                {
-                    this_component->threading_model = ABI::Windows::Foundation::ThreadingType::ThreadingType_BOTH;
-                }
-                else
-                {
-                    return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
-                }
-
-                const WCHAR* xmlns;
-                RETURN_IF_FAILED(xmlReader->MoveToAttributeByName(L"xmlns", nullptr));
-                RETURN_IF_FAILED(xmlReader->GetValue(&xmlns, nullptr));
-                this_component->xmlns = xmlns;
-
-                const WCHAR* activatableClass;
-                RETURN_IF_FAILED(xmlReader->MoveToAttributeByName(L"name", nullptr));
-                RETURN_IF_FAILED(xmlReader->GetValue(&activatableClass, nullptr));
-                g_types[activatableClass] = this_component;
+                RETURN_IF_FAILED(ParseActivatableClassTag(xmlReader, fileName));
+            }
+        }
+        else if (nodeType == XmlNodeType_EndElement)
+        {
+            RETURN_IF_FAILED((xmlReader->GetLocalName(&localName, nullptr)));
+            if (_wcsicmp_l(localName, L"file", locale) == 0)
+            {
+                return S_OK;
+            }
+        }
+    }
+    return S_OK;
+}
+
+HRESULT ParseActivatableClassTag(ComPtr<IXmlReader> xmlReader, LPCWSTR fileName)
+{
+    LPCWSTR localName = nullptr;
+    auto locale = _create_locale(LC_ALL, "C");
+    auto this_component = make_shared<component>();
+    this_component->module_name = fileName;
+    HRESULT hr = xmlReader->MoveToFirstAttribute();
+    // Using this pattern intead of calling multiple MoveToAttributeByName improves performance
+    const WCHAR* activatableClass = nullptr;
+    const WCHAR* threadingModel = nullptr;
+    const WCHAR* xmlns = nullptr;
+    if (S_FALSE == hr)
+    {
+        return HRESULT_FROM_WIN32(ERROR_SXS_MANIFEST_PARSE_ERROR);
+    }
+    else
+    {
+        while (TRUE)
+        {
+            const WCHAR* pwszLocalName;
+            const WCHAR* pwszValue;
+            if (FAILED(hr = xmlReader->GetLocalName(&pwszLocalName, NULL)))
```

Comment:
Consider using `if (FAILED_LOG(...))` which integrates with WIL's error path so you don't need the wprintfs at all.